Add Atomic<T> template

kjbracey opened the pull request:
### Description

Add a C++ `Atomic<T>` template to make atomics even easier. Basically compatible with C++11 `std::atomic<T>`, but using the underlying `core_util_atomic_xxx` functions from mbed_atomic.h, so appropriate for synchronising with interrupts and optimised for uniprocessor.

One extra piece of functionality beyond the `core_util_atomic_xxx` functions is the ability to have an arbitrary atomic type - eg a small structure with 2 `uint16_t`s can be stored in a `uint32_t` container.

Based on `mbed_cxxsupport.h` from #10868.

Requires `MBED_STRUCT_STATIC_ASSERT` fix now in #10837

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Release Notes

* Added `Atomic<T>` template to make atomics easier in C++.

kjbracey committed:
Add Atomic<T> template

Add a C++ `Atomic<T>` template to make atomics even easier. Basically
compatible with C++11 `std::atomic<T>`, but using the underlying
`core_util_atomic_xxx` functions from mbed_atomic.h, so appropriate for
synchronising with interrupts and optimised for uniprocessor.

One extra piece of functionality beyond the `core_util_atomic_xxx`
functions is the ability to have an arbitrary atomic type - eg a small
structure with 2 `uint16_t`s can be stored in a `uint32_t` container.

kegilbert commented on platform/Atomic.h:
```diff
@@ -0,0 +1,1108 @@
+ * - For `bool`, integer types and pointers, storage is compatible with the
+ *   plain types. If necessary, they can be substituted as plain types for C
+ *   compatibility in headers, and accessed using core_util_atomic functions. Eg:
+ *      struct foo {
```
I can't make multi-line suggestions, sorry for the incoming double comment. Wrapping this in @code will resolve doxygen spelling error detection. If you'd rather not wrap this in @code ifdef could be added to the whitelist of words in the spellchecker.

```suggestion
@code
 *      struct foo {
```

kegilbert commented on platform/Atomic.h:
```diff
@@ -0,0 +1,1108 @@
+ *         uint32_t counter;  // Could use core_util_atomic_xxx_u32 from C code, or just have this for structure layout.
+ *      #endif
+ *      };
+ */
```
Part 2 of above comment.
```suggestion
@endcode
 */
```

kjbracey committed:
mbed_atomic_impl.h: Add pre-op (xxx_fetch) forms

Even though C/C++11 don't offer pre-op forms (that do the operation then
return the new value) of their freestanding functions, we will be making
them visible via pre-op operators on `Atomic<T>` (++, --, +=, -=, &=,
|=, ^=).

It is easier to do a pre-op than a post-op, as we can use one
fewer register in the assembler, so it's worth optimising for what will
be quite common cases. Make these forms accessible for `Atomic<T>`, but
don't document them for standalone use at this stage.

kjbracey committed:
Make atomic test use Atomic<T>

Make the atomic functional test use `Atomic<T>` rather than the
standalone functions - this then tests both layers simultaneously.

kjbracey committed:
Test Atomic<T> with user T

kjbracey committed:
mbed_atomic templates: add noexcept

Add noexcept for consistency with upcoming Atomic.h

kjbracey committed:
mbed_atomic.h: Improve template type deduction

Avoid template ambiguities using type_identity_t.

Previously the compiler would be unable to figure out whether

     uint8_t x;
     core_util_atomic_store(&x, 0);

should invoke core_util_atomic_store<uint8_t>, matching the pointer
type, or core_util_atomic_store<int>, matching the value, leading to
an ambiguity error.

Templates now select only on the type of the atomic pointer parameter.

kjbracey committed:
Set unit tests to C++14

Adjust some non-C++11-compatible code that failed as a result.
